common/persistence/versionHistory.go

```diff
@@ -0,0 +1,165 @@
+				currentItem.version,
+				lastItem.version),
+		}
+	}
+
+	if currentItem.eventID <= lastItem.eventID {
+		return &shared.BadRequestError{
+			Message: fmt.Sprintf("cannot add version history with a lower event id %v. Last event id: %v",
+				currentItem.eventID,
+				lastItem.eventID),
+		}
+	}
+
+	if currentItem.version > lastItem.version {
+		// Add a new history
+		v.history = append(v.history, currentItem)
+	} else {
+		// item.version == lastItem.version && item.eventID > lastItem.eventID
+		// Update event  id
+		lastItem.eventID = currentItem.eventID
+	}
+	return nil
+}
+
+// FindLowestCommonVersionHistoryItem returns the lowest version history item with the same version
+func (v *VersionHistory) FindLowestCommonVersionHistoryItem(remote VersionHistory) (VersionHistoryItem, error) {
+	localIdx := len(v.history) - 1
+	remoteIdx := len(remote.history) - 1
+
+	for localIdx >= 0 && remoteIdx >= 0 {
+		localVersionItem := v.history[localIdx]
+		remoteVersionItem := remote.history[remoteIdx]
+		if localVersionItem.version == remoteVersionItem.version {
+			if localVersionItem.eventID > remoteVersionItem.eventID {
+				return remoteVersionItem, nil
+			}
+			return localVersionItem, nil
+		} else if localVersionItem.version > remoteVersionItem.version {
+			localIdx--
+		} else {
+			// localVersionItem.version < remoteVersionItem.version
+			remoteIdx--
+		}
+	}
+	return VersionHistoryItem{}, &shared.BadRequestError{
+		Message: fmt.Sprintf("version history is malformed. No joint point found."),
+	}
+}
+
+// IsAppendable checks if a version history item is appendable
+func (v *VersionHistory) IsAppendable(item VersionHistoryItem) bool {
+	return v.history[len(v.history)-1] == item
+}
+
+// NewVersionHistories initialize new version histories
+func NewVersionHistories(histories []VersionHistory) VersionHistories {
+	if histories == nil || len(histories) == 0 {
+		panic("version histories cannot be empty")
+	}
+	return VersionHistories{
+		versionHistories: histories,
+	}
+}
+
+// FindLowestCommonVersionHistory finds the lowest common version history item among all version histories
+func (h *VersionHistories) FindLowestCommonVersionHistory(history VersionHistory) (VersionHistoryItem, VersionHistory, error) {
+	var versionHistoryItem VersionHistoryItem
+	var versionHistory VersionHistory
+	for _, localHistory := range h.versionHistories {
+		item, err := localHistory.FindLowestCommonVersionHistoryItem(history)
+		if err != nil {
+			return versionHistoryItem, versionHistory, err
+		}
+
+		if item.eventID > versionHistoryItem.eventID {
+			versionHistoryItem = item
+			versionHistory = localHistory
+		}
+	}
+	return versionHistoryItem, versionHistory, nil
+}
+
+// AddHistory add new history into version histories
+// Sample
+func (h *VersionHistories) AddHistory(item VersionHistoryItem, local VersionHistory, remote VersionHistory) error {
+	commonItem := item
+	if local.IsAppendable(commonItem) {
+		for _, historyItem := range remote.history {
```

Comment:
after this for loop, is local == remote?

Reply:
Correct. I will remove this loop. But there will be loop to find the passed in versionHistory in the versionHistories. After the two functions merged, this loop will not be needed